core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -954,7 +954,7 @@ private[spark] class TaskSetManager(
       taskSetExcludelistHelperOpt.foreach(_.updateExcludedForFailedTask(
         info.host, info.executorId, index, failureReason))
       numFailures(index) += 1
-      if (numFailures(index) >= maxTaskFailures) {
+      if (numFailures(index) >= maxTaskFailures && !successful(index)) {
```

Comment:
Thoughts on moving this check to the outside `if` condition ?
Effectively treating a speculative task failure as `countTowardsTaskFailures` = `false`
+CC @Ngone51

Reply:
This seems to break the existing behavior when the 4th failure happens to be `FetchFailed` -  previously it'd abort the stage but now it couldn't.

Reply:
I don't see the potential issues with this change. So I am personally ok with it.  Might be better to add a log for the case.

Reply:
> I don't see the potential issues with this change. So I am personally ok with it. Might be better to add a log for the case.

Thanks for review the pr. I will add some log later.

Reply:
@Ngone51 @mridulm Colud you please take a look when time free? I added some log for this case.

Reply:
I rollback the change about log, because we reset the `numFailures` if some attempt succeed